Use `paste!` to simplify port macros in avr-hal-generic

LuigiPiucco opened the pull request:
Following #449, a draft of the reimplementation. Currently done is the port module and the simple_pwm module. I will be working on the others as well, and feel free to suggest changes/improvements. The intent is to deduplicate the implementations for each CPU, hoping to allow easier refactorings and additions of features. 

Closes #449.

Rahix commented:
Hi, thanks a lot for this work.  You have to be careful with simplifying the interface too much because in some places the chips themselves are not consistent enough to abstract them properly.  I believe this might be the cause for the CI failures here.

In general, my suggestion would be to do this work in one module per PR.  So please open one PR for the `port` module changes and a second one for the `simple_pwm` changes so each can be looked at and worked on in isolation.

LuigiPiucco commented:
> You have to be careful with simplifying the interface too much because in some places the chips themselves are not consistent enough to abstract them properly. I believe this might be the cause for the CI failures here.

Yes, I'm noticing the inconsistencies. For the port module they don't seem that bad, but the PWM one requires extra care, and probably the ones I haven't looked at yet as well. On that note, I should mention I couldn't fit Timer 1 on the Attiny85 into the module. It has a dual waveform and outputs to 2 pins simultaneously, so we probably can't reliably represent that with the current framework. I left it unimplemented with a TODO comment for later. It had an implementation prior, but it considered only one pin, which could lead to issues if the user expects to be able to use its dual normally.

> In general, my suggestion would be to do this work in one module per PR. So please open one PR for the port module changes and a second one for the simple_pwm changes so each can be looked at and worked on in isolation.

Okay, that's a good idea. I'll edit this one to have only the port module as scope, and open others as I proceed to other modules.

LuigiPiucco commented:
I'll mark this as ready for review, since we limited the scope. I rebased and fixed some issues in `arduino-hal` (I forgot to update some board definitions with the new macro). Also built lib for all CPUs and examples for all boards, it should be good now, but let's see if CI catches anything else.

As a side note, and since you thanked me for this, I should say this crate is very high quality, it sure beats the regular Arduino stack by a wide margin. So thank you as well.

Rahix commented on avr-hal-generic/src/port.rs:
```diff
@@ -523,169 +521,176 @@ macro_rules! impl_port_traditional {
     (
-        type Pin = $PinType:ident;
-
         $(#[$pins_attr:meta])*
-        pub struct Pins from $McuPins:ty {
-            $($(#[$pin_attr:meta])* pub $pin:ident: $Pin:ty = $pin_orig:ident,)+
+        pub struct Pins {
+            $($(#[$pin_attr:meta])* pub $pin_name:ident: $pin_type:ty = $pin_orig:ident,)+
+        }
+
+        impl Pins {
+            type Pin = $pin_wrapper:ident;
+            type McuPins = $mcu_pins:ty;
         }
```
Can you comment on the motivation behind this change?  I don't have strong opinions either direction, but would like to hear why you prefer this macro syntax.

LuigiPiucco replied:
This is indeed a more stylistic change. Now that I think about it, I could probably have left this bit untouched, just changing the transcriber. But in general, when writing macros I prefer to keep syntax as close to the base language as possible, and the `Pins from $McuPins` part both introduces a "new" keyword (`from`) and looks more like a syntax for inheritance (in the OOP sense). The type that the macro generates doesn't really inherit the other, in the sense of "class A is a B and therefore can do what B can and be where B would", it's more like they're alternative interfaces. I thought making it seem as an associated type would be more fitting, because 1) it still shows the relationship between the types; 2) it doesn't add any syntax (though inherent associated types is an unstable grammar, I believe, so maybe we could improve); 3) it looks more native to Rust, since it doesn't have proper inheritance.

All that being said, it's mostly superfluous, so if you'd rather make the diff a bit smaller I can change it back.

Rahix replied:
Fair, thanks a lot for the detailed elaboration :)  Sounds good to me!

> All that being said, it's mostly superfluous, so if you'd rather make the diff a bit smaller I can change it back.

I think for the future, it would be a good idea to at least put each of the changes into its own commit as they don't logically relate to one another.  But let's not bother with such perfectionism here, I'll take your changes as they stand right now.

LuigiPiucco replied:
> I think for the future, it would be a good idea to at least put each of the changes into its own commit as they don't logically relate to one another.

I can apply this suggestion, just have a question of preference before: Do you value the ability to compile each commit separately, or having it working by the end of the pull request is enough? There are some instances where using a more micro approach will require introducing changes that would not need to be there otherwise, to maintain self-containment. I'm fine with either, just would rather ask to make sure.

Rahix commented on arduino-hal/src/port/mega.rs:
```diff
@@ -1,340 +1,343 @@
         /// * `PCINT16`: External Interrupt (Pin Change)
         pub a8: atmega_hal::port::PK0 = pk0,
-        /// `A9`
+        /// `PUB A9`
         ///
         /// * `ADC9`: A/D converter input 9
         /// * `PCINT17`: External Interrupt (Pin Change)
         pub a9: atmega_hal::port::PK1 = pk1,
-        /// `A10`
+        /// `PUB A10`
         ///
         /// * `ADC10`: A/D converter input 10
         /// * `PCINT18`: External Interrupt (Pin Change)
         pub a10: atmega_hal::port::PK2 = pk2,
-        /// `A11`
+        /// `PUB A11`
         ///
         /// * `ADC11`: A/D converter input 11
         /// * `PCINT19`: External Interrupt (Pin Change)
         pub a11: atmega_hal::port::PK3 = pk3,
-        /// `A12`
+        /// `PUB A12`
         ///
         /// * `ADC12`: A/D converter input 12
         /// * `PCINT20`: External Interrupt (Pin Change)
         pub a12: atmega_hal::port::PK4 = pk4,
-        /// `A13`
+        /// `PUB A13`
         ///
         /// * `ADC13`: A/D converter input 13
         /// * `PCINT21`: External Interrupt (Pin Change)
         pub a13: atmega_hal::port::PK5 = pk5,
-        /// `A14`
+        /// `PUB A14`
         ///
         /// * `ADC14`: A/D converter input 14
         /// * `PCINT22`: External Interrupt (Pin Change)
         pub a14: atmega_hal::port::PK6 = pk6,
-        /// `A15`
+        /// `PUB A15`
```
Looks like a lot of accidental changes in this file?

LuigiPiucco replied:
Yes, this was the result of carefree regex substitution... 😆
I'm fixing it right away.

Rahix reviewed (approved):
Looks great now, thanks a lot again!
